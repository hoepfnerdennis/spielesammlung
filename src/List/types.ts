interface IBaseGame {
  name: string;
  description: string;
  playersFrom: number;
  playersTo: number;
  age: string;
  duration: string;
  favorite: boolean;
  simpleRules: boolean;
}

export interface IGame extends IBaseGame {
  image: string;
}

enum LinkTypes {
  'Asset',
  'Space',
  'Environment',
  'ContentType',
}

enum SysTypes {
  'Link',
  'Asset',
}

interface IAPIGame extends IBaseGame {
  image?: {
    sys: {
      type: string;
      linkType: LinkTypes;
      id: string;
    };
  };
}

export interface IAsset {
  sys: {
    space: {
      sys: {
        type: string;
        linkType: LinkTypes;
        id: string;
      };
    };
    id: string;
    type: SysTypes;
    createdAt: string;
    updatedAt: string;
    environment: {
      sys: {
        id: string;
        type: SysTypes;
        linkType: LinkTypes;
      };
    };
    revision: number;
    locale: string;
  };
  fields: {
    title: string;
    description: string;
    file: {
      url: string;
      details: {
        size: number;
        image: {
          width: number;
          height: number;
        };
      };
      fileName: string;
      contentType: string;
    };
  };
}

export interface IAPIResponse {
  sys: {
    type: string;
  };
  total: number;
  skip: number;
  limit: number;
  items: [
    {
      sys: {
        space: {
          sys: {
            type: string;
            linkType: LinkTypes;
            id: string;
          };
        };
        id: string;
        type: string;
        createdAt: string;
        updatedAt: string;
        environment: {
          sys: {
            id: string;
            type: string;
            linkType: LinkTypes;
          };
        };
        revision: number;
        contentType: {
          sys: {
            type: string;
            linkType: LinkTypes;
            id: string;
          };
        };
        locale: string;
      };
      fields: IAPIGame;
    }
  ];
  includes?: {
    Asset: IAsset[];
  };
}

/*
{
  "sys": {
    "type": "Array"
  },
  "total": 3,
  "skip": 0,
  "limit": 100,
  
  "includes": {
    "Asset": [
      {
        "sys": {
          "space": {
            "sys": {
              "type": "Link",
              "linkType": "Space",
              "id": "9sxha2f3gm24"
            }
          },
          "id": "2Txvyka47nhfVoh2Sc1OUh",
          "type": "Asset",
          "createdAt": "2020-03-20T14:04:02.201Z",
          "updatedAt": "2020-03-20T14:04:02.201Z",
          "environment": {
            "sys": {
              "id": "master",
              "type": "Link",
              "linkType": "Environment"
            }
          },
          "revision": 1,
          "locale": "en-US"
        },
        "fields": {
          "title": "Mensch ärgere dich nicht",
          "description": "Mensch ärgere dich nicht",
          "file": {
            "url": "//images.ctfassets.net/9sxha2f3gm24/2Txvyka47nhfVoh2Sc1OUh/d277e6dc09f15b5bb5a134ff7082fe4a/photo-1582921017967-79d1cb6702ee",
            "details": {
              "size": 192767,
              "image": {
                "width": 2100,
                "height": 1400
              }
            },
            "fileName": "photo-1582921017967-79d1cb6702ee",
            "contentType": "image/jpeg"
          }
        }
      }
    ]
  }
}
 */
