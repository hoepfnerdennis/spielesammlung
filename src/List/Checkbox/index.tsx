import React from 'react';
import styles from './styles.module.scss';

const Checkbox: React.SFC<{
  label: string;
  onChange: (value: boolean) => void;
  checked: boolean;
}> = ({ label, onChange, checked }): JSX.Element => {
  const handleClick = (): void => {
    onChange(!checked);
  };
  return (
    <button
      type="button"
      className={checked ? styles.checked : styles.unchecked}
      onClick={handleClick}>
      {label}
    </button>
  );
};

export default Checkbox;
