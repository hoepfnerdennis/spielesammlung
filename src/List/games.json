[
    {
        "name": "Mensch ärgere dich nicht",
        "description": "Bei diesem Brettspiel laufen vier Figuren eines jeden Spielers einmal um das Spielfeld. Das Ziel ist dabei die Umrundung und Rückkehr ins eigene Haus. Die Spieler können die gegnerischen Figuren dabei immer wieder aus dem Spiel werfen",
        "players": {
            "from": "2",
            "to": "4"
        },
        "age": "ab 4",
        "duration": "10 - 15"
    },
    {
        "name": "Siedler",
        "description": "Bei diesem Brettspiel laufen vier Figuren eines jeden Spielers einmal um das Spielfeld. Das Ziel ist dabei die Umrundung und Rückkehr ins eigene Haus. Die Spieler können die gegnerischen Figuren dabei immer wieder aus dem Spiel werfen",
        "players": {
            "from": "2",
            "to": "4"
        },
        "age": "ab 8",
        "duration": "10 - 15"
    },
    {
        "name": "Patchwork",
        "description": "Bei diesem Brettspiel laufen vier Figuren eines jeden Spielers einmal um das Spielfeld. Das Ziel ist dabei die Umrundung und Rückkehr ins eigene Haus. Die Spieler können die gegnerischen Figuren dabei immer wieder aus dem Spiel werfen",
        "players": {
            "from": "2",
            "to": "2"
        },
        "age": "ab 12",
        "duration": "10 - 15"
    },
    {
        "name": "Drecksau",
        "description": "Bei diesem Brettspiel laufen vier Figuren eines jeden Spielers einmal um das Spielfeld. Das Ziel ist dabei die Umrundung und Rückkehr ins eigene Haus. Die Spieler können die gegnerischen Figuren dabei immer wieder aus dem Spiel werfen",
        "players": {
            "from": "3",
            "to": "5"
        },
        "age": "ab 4",
        "duration": "10 - 15"
    },
    {
        "name": "Monopoli",
        "description": "Bei diesem Brettspiel laufen vier Figuren eines jeden Spielers einmal um das Spielfeld. Das Ziel ist dabei die Umrundung und Rückkehr ins eigene Haus. Die Spieler können die gegnerischen Figuren dabei immer wieder aus dem Spiel werfen",
        "players": {
            "from": "2",
            "to": "10"
        },
        "age": "ab 8",
        "duration": "10 - 15"
    }
]