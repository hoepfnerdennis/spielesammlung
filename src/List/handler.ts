import { useState, useEffect, useMemo, useCallback } from 'react';
import { IGame, IAPIResponse, IAsset } from './types';

type FilterValues = number[];
type FilterFunction = (value: number) => void;
type SearchFunction = (name: string) => void;
type FilterBoolFunction = (value: boolean) => void;

const order = (a: number, b: number): number => a - b;
const sortByName = (a: IGame, b: IGame): number => {
  const nameA = a.name.toUpperCase();
  const nameB = b.name.toUpperCase();
  if (nameA < nameB) {
    return -1;
  }
  if (nameA > nameB) {
    return 1;
  }
  return 0;
};

const SPACE_ID = '9sxha2f3gm24';
const API_TOKEN = '7LDIC95TsrYOfZwEQnbAuMHtij97kfk5r1dIRiGqT8M';
// For local testing replace BASE_URI with
// const BASE_URI = 'http://localhost:3001/response.json';
const BASE_URI = `https://cdn.contentful.com/spaces/${SPACE_ID}/environments/master/entries?access_token=${API_TOKEN}`;

const useGames = (): [
  IGame[],
  FilterValues,
  FilterFunction,
  FilterValues,
  FilterFunction,
  SearchFunction,
  boolean | undefined,
  FilterBoolFunction
] => {
  const [games, setGames] = useState<IGame[]>([]);

  const [filterByPlayersFromValue, setFilterByPlayersFromValue] = useState<number | undefined>();
  const [filterByPlayersToValue, setFilterByPlayersToValue] = useState<number | undefined>();
  const [filterByFavoriteValue, setFilterByFavoriteValue] = useState<boolean | undefined>();
  const [searchTerm, setSearchTerm] = useState<string | undefined>();

  const mapResultsToGames = (results: IAPIResponse): IGame[] => {
    const findImageForGame = (assets: IAsset[], id: string): string => {
      return assets.find(asset => asset.sys.id === id)?.fields.file.url || '';
    };
    return results.items
      .map(item => {
        const {
          age,
          description,
          duration,
          name,
          playersFrom,
          playersTo,
          favorite,
          simpleRules,
          image: imageRef,
        } = item.fields;
        const gameItem: IGame = {
          age,
          description,
          duration,
          name,
          playersFrom,
          playersTo,
          favorite,
          simpleRules,
          image: '',
        };
        if (imageRef && results.includes?.Asset) {
          const image: string = findImageForGame(results.includes.Asset, imageRef.sys.id);
          gameItem.image = image;
        }
        return gameItem;
      })
      .sort(sortByName);
  };

  const loadGames = useCallback(async (url: string): Promise<void> => {
    try {
      const response = await fetch(url);
      const data: IAPIResponse = await response.json();
      const gamesFromAPI: IGame[] = mapResultsToGames(data);
      setGames(gamesFromAPI);
    } catch {
      setGames([]);
    }
  }, []);

  useEffect(() => {
    loadGames(BASE_URI);
  }, [loadGames]);

  const performSearch = (
    from: number | undefined,
    to: number | undefined,
    term: string | undefined,
    favorite: boolean | undefined
  ): void => {
    let url = `${BASE_URI}&content_type=game`;
    if (from) {
      url += `&fields.playersFrom[lte]=${from}`; // &fields.playersTo[gte]=${from}`;
    }
    if (to) {
      url += `&fields.playersTo[gte]=${to}`; // &fields.playersFrom[lte]=${to}`;
    }
    if (term) {
      url += `&fields.name[match]=${term}`;
    }
    if (favorite) {
      url += `&fields.favorite=${favorite}`;
    }
    loadGames(url);
  };

  const playersFromValues: number[] = useMemo(() => {
    return games
      .map(game => game.playersFrom)
      .filter((players, index, array) => array.indexOf(players) === index)
      .sort(order);
  }, [games]);

  const playersToValues: number[] = useMemo(() => {
    return games
      .map(game => game.playersTo)
      .filter((players, index, array) => array.indexOf(players) === index)
      .sort(order);
  }, [games]);

  const filterByPlayersFrom = (value: number): void => {
    if (value) {
      setFilterByPlayersFromValue(value);
      performSearch(value, filterByPlayersToValue, searchTerm, filterByFavoriteValue);
    } else {
      setFilterByPlayersFromValue(undefined);
      performSearch(undefined, filterByPlayersToValue, searchTerm, filterByFavoriteValue);
    }
  };

  const filterByPlayersTo = (value: number): void => {
    if (value) {
      setFilterByPlayersToValue(value);
      performSearch(filterByPlayersFromValue, value, searchTerm, filterByFavoriteValue);
    } else {
      setFilterByPlayersToValue(undefined);
      performSearch(filterByPlayersFromValue, undefined, searchTerm, filterByFavoriteValue);
    }
  };

  const searchForName = (term: string): void => {
    if (term?.length > 1) {
      const lowerTerm = term.toLowerCase();
      setSearchTerm(lowerTerm);
      performSearch(
        filterByPlayersFromValue,
        filterByPlayersToValue,
        lowerTerm,
        filterByFavoriteValue
      );
    } else {
      setSearchTerm(undefined);
      performSearch(
        filterByPlayersFromValue,
        filterByPlayersToValue,
        undefined,
        filterByFavoriteValue
      );
    }
  };

  const filterByFavorite = (value: boolean): void => {
    if (value === true) {
      setFilterByFavoriteValue(value);
      performSearch(filterByPlayersFromValue, filterByPlayersToValue, searchTerm, value);
    } else {
      setFilterByFavoriteValue(undefined);
      performSearch(filterByPlayersFromValue, filterByPlayersToValue, searchTerm, undefined);
    }
  };

  return [
    games,
    playersFromValues,
    filterByPlayersFrom,
    playersToValues,
    filterByPlayersTo,
    searchForName,
    filterByFavoriteValue,
    filterByFavorite,
  ];
};

export default useGames;
