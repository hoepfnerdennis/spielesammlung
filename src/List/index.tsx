import React, { Suspense, useContext } from 'react';
import { IGame } from '../Store/types';
import Filters from '../Filters';
import styles from './styles.module.css';
import DataContext from '../Store/DataContext';

const Entry = React.lazy(() => import(/* webpackChunkName: "entry" */ '../Entry'));

const List: React.SFC = (): JSX.Element => {
  const { games } = useContext(DataContext);

  return (
    <>
      <div className={styles.filterContainer}>
        <span>{games.length} Spiele</span>
        <Filters />
      </div>
      {games.map(
        ({
          id,
          age,
          description,
          duration,
          name,
          playersFrom,
          playersTo,
          favorite,
          simpleRules,
          image,
          drinkingGame,
        }: IGame) => (
          <Suspense fallback={null} key={id}>
            <Entry
              id={id}
              age={age}
              description={description}
              duration={duration}
              name={name}
              playersFrom={playersFrom}
              playersTo={playersTo}
              favorite={favorite}
              simpleRules={simpleRules}
              image={image}
              drinkingGame={drinkingGame}
            />
          </Suspense>
        )
      )}
    </>
  );
};

export default List;
