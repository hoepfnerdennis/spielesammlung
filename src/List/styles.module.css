.filters {
    display: flex;
    flex-direction: row;
    justify-content: space-between;

    max-width: 800px;
    margin: 1rem auto;
}