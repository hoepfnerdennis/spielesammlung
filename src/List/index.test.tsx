import React from 'react';
import { render } from '@testing-library/react';
import List from './index';
import getGame from '../../mockData/test-games';

describe('List', () => {
  it('should render and match snapshot', () => {
    const games = [getGame()];
    const { container } = render(
      <List games={games} activeFilters={new Map()} setFilter={jest.fn()} />
    );
    expect(container).toMatchSnapshot();
  });
});
