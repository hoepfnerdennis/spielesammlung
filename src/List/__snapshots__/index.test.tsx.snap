// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render and match snapshot 1`] = `
<div>
  <div
    class="filterContainer"
  >
    <span>
      1
       Spiele
    </span>
    <button
      class="button"
      type="button"
    >
      Filter anzeigen
    </button>
  </div>
</div>
`;
